docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+
+Please note that the practical consequence of focusing on in-place rollout strategies, is that the possibility to implement different types of custom rollout strategies, even if technically possible, won’t be validated in this first iteration (future goal).
+
+Another important point to surface, before digging into implementation details of the proposal, is the fact that this proposal is not tackling the problem of improving CAPI API to embrace all the possibilities that rollout extensions are introducing.
+
+E.g. If a rollout extension introduces support for in-place updates, using “BootstrapConfig” (emphasis on bootstrap) as the place where most of the machine configurations are defined seems not ideal.
+
+However, at the same time we would like to make it possible for Cluster API users to start exploring this field, gain experience, and report back so we can have concrete use cases and real-world feedback to evolve our API.
+
+### Tenets
+
+#### Same UX
+
+Cluster API user experience MUST be the same when using default, immutable updates or when using rollout extensions: e.g. in order to trigger a MachineDeployment rollout, you have to rotate a template, etc.
+
+#### Fallback to Immutable rollouts
+
+If rollout extensions can not cover the totality of the desired changes, users SHOULD be able to defer to Cluster API’s default, immutable rollouts. This is important for a couple of reasons:
+
+* It allows to implement custom rollout strategies incrementally, without the need to cover all use cases up-front.
+* There are case when replacing the machine will always be necessary:
+    * When it is not possible to recover the machine, e.g. hardware failure.
+    * When the user determines that recovering the machine is too complex/costly vs replacing it. 
+    * Automatic machine remediation (unless you use external remediation strategies)
+
+#### Clean separation of concern
+
+The rollout extension will be responsible to perform the updates on a single machine.
+
+The responsibility to determine which machine should be rolled out as well as the responsibility to handle rollout options like MaxSurge/MaxUnavailable will remain on the controllers owning the machine (e.g. KCP, MD controller).
+
+### Goals
+
+- Enable the implementation of in-place update strategies.
+- Allow users to update Kubernetes clusters using pluggable External Update Extension.
+- Maintain a coherent user experience for both rolling and in-place updates.
+- Support External Update Extensions for both Control Plane (KCP or others) and MachineDeployment controlled machines.
+- Allow in-place updates for single-node clusters without the requirement to reprovision hosts.
+
+### Non-Goals
+
+- To provide rollbacks in case of an in-place update failure. Failed updates need to be fixed manually by the user on the machine or by replacing the machine.
```

Comment:
Controller will never rollout the machine in case of in-place update failure. At most, MHC might mark the machine for remediation. But that's a separate process.

Reply:
Yes, the idea is for both the update fallback strategy and MHC remediation (already is) to be optional